# Week 5 Lab — Advanced Python + NumPy

**AUCA · Introduction to Big Data Analytics**

**Name:** NIYITANGA DENYS  
**Student ID:** 29125  
**Program:** Software Engeneering

**Dataset:** Week 5 Kigali Weather — 60 daily records, September To October 2026.

This notebook completes Lab 3 using the Week 5 lecture methods: list comprehensions, `lambda`, NumPy arrays, vectorized operations, boolean masks, and anomaly detection made by Niyitanga Denys Bahati.

## Load the Kigali weather data

In [ ]:
import numpy as np
import time

with open("data/week5_kigali_weather.csv") as f:
    lines = f.readlines()[1:]

rows = [ln.strip().split(",") for ln in lines]
dates = np.array([r[1] for r in rows])
temps = np.array([float(r[2]) for r in rows])
rain = np.array([float(r[3]) for r in rows])
hum = np.array([float(r[4]) for r in rows])

print(f"Loaded {temps.size} days.")
print(f"Temperature mean: {temps.mean():.2f} °C")

Loaded 60 days.
Temperature mean: 24.00 °C


## Part 1 · List Comprehensions

### Exercise 1.1 — Transform

In [ ]:
temps_part1 = [24.4, 25.9, 28.6, 19.3, 26.0, 22.1]
temps_f = [t * 9/5 + 32 for t in temps_part1]
print(f"Fahrenheit: {temps_f}.:2f")

Fahrenheit: [75.92, 78.62, 83.48, 66.74000000000001, 78.8, 71.78].:2f


### Exercise 1.2 — Filter

In [ ]:
hot = [t for t in temps_part1 if t >= 26]
cool_rounded = [round(t) for t in temps_part1 if t < 24]
print(f"Hot: {hot}")
print(f"Cool rounded: {cool_rounded}")

Hot: [28.6, 26.0]
Cool rounded: [19, 22]


### Exercise 1.3 — Convert types

In [ ]:
raw = ["24.4", "25.9", "28.6", "19.3"]
raw_floats = [float(t) for t in raw]
print(f"Floats: {raw_floats}")
print(f"Average: {sum(raw_floats) / len(raw_floats)}")

Floats: [24.4, 25.9, 28.6, 19.3]
Average: 24.55


## Part 2 · lambda 

### Exercise 2.1 — Rank with `sorted`

In [ ]:
results = [("Aline", 85), ("Eric", 78), ("Diane", 91), ("Samuel", 92), ("Grace", 76)]
ranked = sorted(results, key=lambda x: x[1], reverse=True)
print(f"Top 3: {ranked[:3]}")

Top 3: [('Samuel', 92), ('Diane', 91), ('Aline', 85)]


### Exercise 2.2 — max with a key

In [ ]:
rain_log = [(3, 0.0), (7, 12.5), (11, 3.2), (14, 26.2), (18, 8.9)]
wettest_day = max(rain_log, key=lambda x: x[1])
print(f"Wettest day: {wettest_day}")

Wettest day: (14, 26.2)


## Part 3 · NumPy Arrays 

### Exercise 3.1 — The speed test
The exact speedup depends on the computer running the notebook. The Week 5 notes describe approximately 50× as a typical result and explain that 30×–200× can be normal.

In [ ]:
numbers = list(range(1_000_000))
arr = np.array(numbers)

t0 = time.time()
doubled_python = [n * 2 for n in numbers]
python_time = time.time() - t0

t0 = time.time()
doubled_numpy = arr * 2
numpy_time = time.time() - t0

speedup = python_time / numpy_time if numpy_time > 0 else float("inf")
print(f"Python: {python_time:.6f} s")
print(f"NumPy:  {numpy_time:.6f} s")
print(f"NumPy speedup: about {speedup:.1f}×")

Python: 0.118510 s
NumPy:  0.019447 s
NumPy speedup: about 6.1×


### Exercise 3.2 — Describe the temperatures

In [ ]:
print(f"Mean: {temps.mean():.2f} °C")
print(f"Max: {temps.max():.2f} °C")
print(f"Min: {temps.min():.2f} °C")
print(f"Std: {temps.std():.2f} °C")

Mean: 24.00 °C
Max: 28.60 °C
Min: 19.30 °C
Std: 2.61 °C


### Exercise 3.3 — The hottest day

In [ ]:
position = np.argmax(temps)
print(f"Hottest day: {dates[position]}")
print(f"Temperature: {temps[position]:.1f} °C")

Hottest day: 2026-09-17
Temperature: 28.6 °C


## Part 4 · Boolean Masking 

### Exercise 4.1 — Hot days

In [ ]:
hot_mask = temps >= 26
print(f"Number of hot days: {hot_mask.sum()}")
print(f"Average hot-day temperature: {temps[hot_mask].mean():.2f} °C")

Number of hot days: 18
Average hot-day temperature: 27.11 °C


### Exercise 4.2 — Rain report

In [ ]:
rain_mask = rain > 0
print(f"Rainy days: {rain_mask.sum()}")
print(f"Total rainfall: {rain.sum():.1f} mm")
print(f"Heaviest single-day rainfall: {rain.max():.1f} mm")

Rainy days: 16
Total rainfall: 331.2 mm
Heaviest single-day rainfall: 26.2 mm


### Exercise 4.3 — Cross-array analysis

In [ ]:
print(f"Average temperature on rainy days: {temps[rain > 0].mean():.2f} °C")
print(f"Average temperature on dry days: {temps[rain == 0].mean():.2f} °C")
hot_and_rainy = (temps >= 26) & (rain > 0)
print(f"Hot AND rainy days: {hot_and_rainy.sum()}")
print(f"Days with humidity >= 70%: {(hum >= 70).sum()}")

Average temperature on rainy days: 24.14 °C
Average temperature on dry days: 23.95 °C
Hot AND rainy days: 5
Days with humidity >= 70%: 16


## Part 5 · Anomalies & Trends 

### Exercise 5.1 — Anomaly detection

In [ ]:
anomaly = temps - temps.mean()
above_mean_mask = anomaly > 3
below_mean_mask = anomaly < -3

print("Dates more than 3 °C above the mean:")
print(dates[above_mean_mask])
print(f"Days more than 3 °C below the mean: {below_mean_mask.sum()}")

Dates more than 3 °C above the mean:
['2026-09-07' '2026-09-09' '2026-09-11' '2026-09-12' '2026-09-14'
 '2026-09-16' '2026-09-17' '2026-09-19' '2026-09-22']
Days more than 3 °C below the mean: 8


### Exercise 5.2 — Is the season changing?

In [ ]:
first30_avg = temps[:30].mean()
last30_avg = temps[30:].mean()
print(f"First 30 days average: {first30_avg:.2f} °C")
print(f"Last 30 days average: {last30_avg:.2f} °C")
# Kigali got cooler because the last 30-day average is lower than the first 30-day average.

First 30 days average: 26.21 °C
Last 30 days average: 21.78 °C


## Bonus — Top 5 wettest days

In [ ]:
wettest = sorted(zip(dates, rain), key=lambda x: x[1], reverse=True)
print("Top 5 wettest days:")
print(*wettest[:5], sep="\n")

Top 5 wettest days:
(np.str_('2026-09-07'), np.float64(26.2))
(np.str_('2026-09-14'), np.float64(25.1))
(np.str_('2026-09-21'), np.float64(24.0))
(np.str_('2026-09-05'), np.float64(22.9))
(np.str_('2026-09-28'), np.float64(22.9))


## Part 6 · Reflection

**1. What did NumPy hide from you, and why was the loop version important first?**  
NumPy hides the repeated iteration, accumulation, and element-by-element bookkeeping behind vectorized operations such as `temps.mean()`, `temps.max()`, and boolean masks. Learning the loop version first was important because it made the underlying process understandable before using the shorter and faster NumPy operations.

**2. One insight about Kigali's weather:**  
The 60-day data shows a clear cooling trend between the two halves of the period. The first 30 days averaged **26.21°C**, while the last 30 days averaged **21.78°C**, a decrease of about **4.43°C**. Rainy days were only slightly warmer on average than dry days (24.14°C vs 23.95°C).